GreedyParameterPointSelector.C

```diff
@@ -0,0 +1,465 @@
+        for (int i = 0; i < param_space_size; i++) {
+            vec.item(0) = param_space_min + (i + 1) * frequency;
+            parameter_points_vec.push_back(vec);
+        }
+    }
+    else
+    {
+        double frequency = abs(param_space_max - param_space_min) / param_space_size;
+        for (int i = 0; i < param_space_size; i++) {
+            vec.item(0) = param_space_min + i * frequency;
+            parameter_points_vec.push_back(vec);
+        }
+    }
+
+    constructObject(parameter_points_vec, tolerance, saturation,
+        subset_size, convergence_subset_size, use_centroid, random_seed, debug_algorithm);
+}
+
+void
+GreedyParameterPointSelector::constructObject(
+    std::vector<Vector> parameter_points,
+    double tolerance,
+    double saturation,
+    int subset_size,
+    int convergence_subset_size,
+    bool use_centroid,
+    int random_seed,
+    bool debug_algorithm)
+{
+    CAROM_VERIFY(parameter_points.size() > 0);
+    CAROM_VERIFY(tolerance > 0.0);
+    CAROM_VERIFY(saturation > 0.0);
+    CAROM_VERIFY(subset_size > 0 && subset_size <= parameter_points.size());
+    CAROM_VERIFY(convergence_subset_size > 0 && convergence_subset_size <= parameter_points.size());
+    CAROM_VERIFY(random_seed > 0);
+
+    d_parameter_points = parameter_points;
+    d_tol = tolerance;
+    d_sat = saturation;
+    d_subset_size = subset_size;
+    d_convergence_subset_size = convergence_subset_size;
+    d_use_centroid = use_centroid;
+    d_max_error = 0;
+    d_next_point_to_sample = -1;
+    d_next_point_requiring_residual = -1;
+    d_point_requiring_residual_computed = false;
+    d_iteration_started = false;
+    d_subset_created = false;
+    d_procedure_completed = false;
+    d_subset_counter = 0;
+    d_counter = -1;
+    d_debug_algorithm = debug_algorithm;
+
+    for (int i = 0; i < d_parameter_points.size() - 1; i++) {
+        CAROM_VERIFY(d_parameter_points[i].dim() == d_parameter_points[i + 1].dim());
+    }
+
+    for (int i = 0 ; i < d_parameter_points.size(); i++) {
+        d_parameter_point_errors.push_back(INT_MAX);
+        d_parameter_point_random_indices.push_back(i);
+    }
+    rng.seed(random_seed);
+}
+
+int
+GreedyParameterPointSelector::getNextParameterPoint()
+{
+    if (d_parameter_sampled_indices.size() == d_parameter_points.size())
+    {
+        d_procedure_completed = true;
+        return -1;
+    }
+    if (d_iteration_started || d_procedure_completed)
+    {
+        return -1;
+    }
+
+    if (d_parameter_sampled_indices.size() == 0)
+    {
+        // get center-most point
+        // obtain the centroid and find the point closest to centroid
+        if (d_use_centroid)
+        {
+            Vector centroid(d_parameter_points[0]);
+            for (int i = 1; i < d_parameter_points.size(); i++) {
+                centroid += d_parameter_points[i];
+            }
+            centroid.mult(1.0 / d_parameter_points.size(), centroid);
+
+            double min_dist_to_centroid = INT_MAX;
```

Comment:
This should work in all practical cases, but it would be a bug if the minimum distance is greater than `INT_MAX` (obviously should not happen). It would be safer to initialize this to the distance for `i==0`.